lms/djangoapps/program_enrollments/tests/test_signals.py

```diff
@@ -0,0 +1,193 @@
+
+        # update will create a second record
+        self.provider_config.organization = None
+        self.provider_config.save()
+        self.assertEqual(len(SAMLProviderConfig.objects.all()), 2)
+
+        UserSocialAuth.objects.create(
+            user=self.user,
+            uid='{0}:{1}'.format(self.provider_slug, self.external_id)
+        )
+        program_enrollment.refresh_from_db()
+        self.assertIsNone(program_enrollment.user)
+
+    def test_learner_already_enrolled_in_course(self):
+        course_key = self.course_keys[0]
+        course = CourseOverview.objects.get(id=course_key)
+        CourseEnrollmentFactory(user=self.user, course=course, mode=CourseMode.VERIFIED)
+
+        program_enrollment = self._create_waiting_program_enrollment()
+        program_course_enrollments = self._create_waiting_course_enrollments(program_enrollment)
+
+        UserSocialAuth.objects.create(
+            user=self.user,
+            uid='{0}:{1}'.format(self.provider_slug, self.external_id)
+        )
+
+        self._assert_program_enrollment_user(program_enrollment)
+
+        duplicate_program_course_enrollment = program_course_enrollments[0]
+        self._assert_program_course_enrollment(duplicate_program_course_enrollment, CourseMode.VERIFIED)
+
+        program_course_enrollment = program_course_enrollments[1]
+        self._assert_program_course_enrollment(program_course_enrollment)
+
+    def test_enrolled_with_no_course_enrollments(self):
+        program_enrollment = self._create_waiting_program_enrollment()
+
+        UserSocialAuth.objects.create(
+            user=self.user,
+            uid='{0}:{1}'.format(self.provider_slug, self.external_id)
+        )
+
+        self._assert_program_enrollment_user(program_enrollment)
+
+    def test_create_social_auth_with_no_waiting_enrollments(self):
+        """
+        No exceptions should be raised if there are no enrollments to update
+        """
+        UserSocialAuth.objects.create(
+            user=self.user,
+            uid='{0}:{1}'.format(self.provider_slug, self.external_id)
+        )
+
+    def test_create_social_auth_provider_has_no_organization(self):
+        """
+        No exceptions should be raised if provider is not linked to an organization
+        """
+        provider = SAMLProviderConfigFactory.create()
+        UserSocialAuth.objects.create(
+            user=self.user,
+            uid='{0}:{1}'.format(provider.slug, self.external_id)
+        )
+
+    def test_create_social_auth_non_saml_provider(self):
+        """
+        No exceptions should be raised for a non-SAML uid or if a SAML provider cannot be found
+        """
+        UserSocialAuth.objects.create(
+            user=self.user,
+            uid='google-oauth-user@gmail.com'
+        )
+        UserSocialAuth.objects.create(
+            user=self.user,
+            uid='123:123:123'
+        )
+
+    def test_log_on_enrollment_failure(self):
```

Comment:
Please write a test case for other exceptions to ensure that is logged. See my previous comment on logging for general error case

Reply:
I've added test cases for the other exceptions minus one.  `SAMLProviderConfig.MultipleObjectsReturned` is actually not possible now that we're using the ConfigurationModel function to filter.  I added it for sanity jic something changes elsewhere since the side effect here is non-obvious.

Reply:
Got it. Make sense